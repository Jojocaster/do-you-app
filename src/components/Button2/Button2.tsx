import { MaterialCommunityIcons } from '@expo/vector-icons'
import React, { useRef } from 'react'
import { Animated, StyleSheet, TouchableWithoutFeedback } from 'react-native'
import Colors from '../../constants/Colors'
import useColorScheme from '../../hooks/useColorScheme'
import { Text, View } from '../Themed'

interface ButtonProps {
  children?: string
  elevated?: boolean
  iconSize?: number
  onPress?: () => void
  icon?: React.ComponentProps<typeof MaterialCommunityIcons>['name']
  variant?: 'xs' | 'sm' | 'md' | 'lg'
}

const Content: React.FC<
  Pick<ButtonProps, 'children' | 'icon' | 'variant' | 'iconSize'> & {
    visible?: boolean
  }
> = ({ children, icon, visible = true, variant = 'md', iconSize }) => {
  const theme = useColorScheme()
  const activeVariant = variants[variant]

  return (
    <View style={[styles.content, { opacity: visible ? 1 : 0 }]}>
      {icon && (
        <MaterialCommunityIcons
          name={icon}
          size={iconSize || activeVariant.text.fontSize}
          style={{
            marginRight: children ? 3 : 0,
            color: Colors[theme].button.text,
          }}
        />
      )}
      {children && (
        <Text
          style={[
            styles.text,
            activeVariant.text,
            { color: Colors[theme].button.text },
          ]}
        >
          {children}
        </Text>
      )}
    </View>
  )
}

export const Button2: React.FC<ButtonProps> = (props) => {
  const scale = useRef(new Animated.Value(1))
  const theme = useColorScheme()
  const { elevated, onPress, variant = 'md' } = props

  const activeVariant = variants[variant]

  const onPressIn = () => {
    Animated.spring(scale.current, {
      toValue: 0.8,
      friction: 4,
      useNativeDriver: true,
    }).start()
  }

  const onPressOut = () => {
    Animated.spring(scale.current, {
      toValue: 1,
      friction: 2,
      useNativeDriver: true,
    }).start()
  }

  return (
    <TouchableWithoutFeedback
      delayLongPress={0}
      onPress={onPress}
      onPressIn={onPress ? onPressIn : undefined}
      onPressOut={onPress ? onPressOut : undefined}
    >
      <Animated.View
        style={[styles.container, elevated ? styles.elevated : null]}
      >
        {/* TODO: clean up */}
        <Animated.View
          style={[
            styles.button,
            activeVariant.button,
            {
              transform: [{ scale: scale.current }],
              backgroundColor: Colors[theme].button.back,
            },
          ]}
        >
          <Content {...props} visible={false} />
          <View
            style={[
              styles.button,
              activeVariant.button,
              styles.inner,
              { backgroundColor: Colors[theme].button.background },
            ]}
          >
            <Content {...props} />
          </View>
        </Animated.View>
      </Animated.View>
    </TouchableWithoutFeedback>
  )
}

const xsStyles = StyleSheet.create({
  button: {
    paddingHorizontal: 8,
  },
  text: {
    fontSize: 8,
  },
})

const smStyles = StyleSheet.create({
  button: {
    paddingHorizontal: 8,
  },
  text: {
    fontSize: 10,
  },
})

const mdStyles = StyleSheet.create({
  button: {
    paddingHorizontal: 16,
  },
  text: {
    fontSize: 12,
  },
})

const lgStyles = StyleSheet.create({
  button: {
    paddingHorizontal: 16,
    paddingVertical: 6,
  },
  text: {
    fontSize: 13,
  },
})

const variants = {
  xs: xsStyles,
  sm: smStyles,
  md: mdStyles,
  lg: lgStyles,
}

const styles = StyleSheet.create({
  container: {
    position: 'relative',
    display: 'flex',
    paddingLeft: 4,
    paddingTop: 4,
    flexWrap: 'wrap',
  },
  button: {
    backgroundColor: 'white',
    borderRadius: 3,
    borderWidth: 1,
    borderBottomColor: 'black',
    paddingVertical: 4,
  },
  elevated: {
    elevation: 2,
    shadowOffset: {
      height: 3,
      width: 3,
    },
    shadowOpacity: 0.1,
  },
  // buttonSm: {
  //   paddingHorizontal: 8,
  // },
  // buttonLg: {
  //   paddingHorizontal: 16,
  // },
  inner: {
    position: 'absolute',
    top: -4,
    left: -5,
  },
  content: {
    backgroundColor: 'transparent',
    display: 'flex',
    alignItems: 'center',
    flexDirection: 'row',
  },
  text: {
    color: 'black',
    fontWeight: 'bold',
    textTransform: 'uppercase',
  },
  // textSm: {
  //   fontSize: 10,
  // },
  // textLg: {
  //   fontSize: 12,
  // },
})
