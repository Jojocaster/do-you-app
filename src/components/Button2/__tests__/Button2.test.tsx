import { Button2 } from '../Button2'
import renderer, { act } from 'react-test-renderer'
import { fireEvent } from '@testing-library/react-native'
import { Animated, Pressable } from 'react-native'
import { MaterialCommunityIcons } from '@expo/vector-icons'
import * as RN from 'react-native'

const mockOnPress = jest.fn()

const mockStore = {
  settings: {
    darkTheme: false,
  },
  tracksInfo: { loading: true, tracks: [] },
}

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn((cb) => cb(mockStore)),
}))

describe('Button', () => {
  it('should render', () => {
    const wrapper = renderer.create(<Button2>foo</Button2>)
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  describe('when `onPress` is defined', () => {
    it('should be called on press', () => {
      const wrapper = renderer.create(
        <Button2 onPress={mockOnPress}>foo</Button2>
      )
      const button = wrapper.root.findByType(RN.TouchableWithoutFeedback)
      fireEvent.press(button)
      expect(mockOnPress).toHaveBeenCalled()
    })

    // it('should call animated on pressIn & pressOut', async () => {
    //   const wrapper = renderer.create(
    //     <Button2 onPress={mockOnPress}>foo</Button2>
    //   )
    //   const button = wrapper.root.findByType(Pressable)
    //   act(() => {
    //     fireEvent(button, 'pressIn')
    //     fireEvent(button, 'pressOut')
    //   })
    //   act(() => {
    //     expect(button.props.onPressIn).toHaveBeenCalledTimes(1)
    //   })
    // })
  })

  describe('when `icon` is defined', () => {
    it('should display the right icon', () => {
      const wrapper = renderer.create(
        <Button2 icon="ab-testing" onPress={mockOnPress}>
          foo
        </Button2>
      )
      expect(wrapper.root.findAllByType(MaterialCommunityIcons)).toBeDefined()
    })
  })
})
