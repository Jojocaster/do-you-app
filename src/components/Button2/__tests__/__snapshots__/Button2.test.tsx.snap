// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render 1`] = `
<View
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "display": "flex",
      "flexWrap": "wrap",
      "paddingLeft": 4,
      "paddingTop": 4,
      "position": "relative",
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "white",
        "borderBottomColor": "black",
        "borderRadius": 3,
        "borderWidth": 1,
        "paddingHorizontal": 16,
        "paddingVertical": 4,
        "transform": Array [
          Object {
            "scale": 1,
          },
        ],
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#fff",
          },
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "display": "flex",
              "flexDirection": "row",
            },
            Object {
              "opacity": 0,
            },
          ],
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000",
            },
            Array [
              Object {
                "color": "black",
                "fontWeight": "bold",
                "textTransform": "uppercase",
              },
              Object {
                "fontSize": 12,
              },
            ],
          ]
        }
      >
        foo
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#fff",
          },
          Array [
            Object {
              "backgroundColor": "white",
              "borderBottomColor": "black",
              "borderRadius": 3,
              "borderWidth": 1,
              "paddingVertical": 4,
            },
            Object {
              "paddingHorizontal": 16,
            },
            Object {
              "left": -5,
              "position": "absolute",
              "top": -4,
            },
            Object {
              "backgroundColor": "#F3DD7C",
            },
          ],
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
            },
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "display": "flex",
                "flexDirection": "row",
              },
              Object {
                "opacity": 1,
              },
            ],
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000",
              },
              Array [
                Object {
                  "color": "black",
                  "fontWeight": "bold",
                  "textTransform": "uppercase",
                },
                Object {
                  "fontSize": 12,
                },
              ],
            ]
          }
        >
          foo
        </Text>
      </View>
    </View>
  </View>
</View>
`;
